client/internal/wsreceiver.go

```diff
@@ -43,22 +43,39 @@ func NewWSReceiver(
 
 // ReceiverLoop runs the receiver loop. To stop the receiver cancel the context.
 func (r *wsReceiver) ReceiverLoop(ctx context.Context) {
-	runContext, cancelFunc := context.WithCancel(ctx)
-
-out:
 	for {
-		var message protobufs.ServerToAgent
-		if err := r.receiveMessage(&message); err != nil {
-			if ctx.Err() == nil && !websocket.IsCloseError(err, websocket.CloseNormalClosure) {
-				r.logger.Errorf("Unexpected error while receiving: %v", err)
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			result := make(chan struct {
+				message *protobufs.ServerToAgent
+				err     error
+			}, 1)
+
+			go func() {
```

Comment:
Can we avoid creating a goroutine for every message received?

Probably not a big deal for the client, we don't receive that many messages for this to have a major performance impact.

If there is an easy way to avoid it it would be nice, otherwise probably don't bother, this is ok.

Reply:
I couldn't think of a clean way to do it; Happy to update if you have suggestions.